### DISCUSSION TITLE
Add Attributes to the generated property

### DISCUSSION THREAD

**Discussion by hutterm** [canton7/PropertyChanged.SourceGenerator#21] [Ideas] [answered]
first of all, great library!!!

Is there a way to put an Attribute on the generated property without having to go through `MetadataTypeAttribute` on the class?

**Comment by canton7** [comment]
Thanks! No, there's no support for that currently. It might be worth adding though! What sort of attributes specifically are you thinking of?

Do bear in mind, those attributes will have to be applicable to both fields and properties, and the tooling which uses them will have to know to ignore the attribute that was added to the field.

**Comment by hutterm** [comment]
hmm, I actually didn't consider that a specific attribute might only be applicable to a property only and not a field.
I just tried wrapping a attribute with another but that doesn't work as attributes only allow simple parameters.
I guess this might not be as easy as I first imagined. just copying the attribute might have unwanted implications

... one more reason to get partial properties in c#

**Reply by canton7** [reply]
Yeah, it's a nasty problem (which partial properties would solve very  nicely!). Another option is to support a magic comment containing the attribute, but then you don't get any design-time help writing it.

**Reply by hutterm** [reply]
magic comment doesn't really sound right either...
I guess I would keep to MetadataTypeAttribute for now

**Reply by hutterm** [reply]
@canton7 huh, turns out, I actually do really need a way apart from `MetadataTypeAttribute`
So until there's some other way, I'd actually appreciate the magic comment option!

**Reply by canton7** [reply]
Thinking about this a bit, I can come up with a few different ways of doing this.

First, a magic comment. Something like:

```cs
// INPC: [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
[Notify] private int _foo;
```

This would be copied verbatim into the generated code. The downside is that the user needs to fully-qualify all types (as the generated file doesn't have any 'using' statements), and there's no design-time checking that the syntax is valid. It's also slightly painful to find all related comments in Roslyn, but it isn't too bad.

Another is to use an attribute, like:

```cs
[PropertyAttribute("System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)")]
```

This just removes the whole "magic comment" thing and makes the usage discoverable (someone can mouseover the PropertyAttribute to read the docs for it), and also makes it easier to process. It has the same downsides around fully-qualifying types and design-time checking though.

We could take this a step further:

```cs
[PropertyAttribute(typeof(EditorBrowsableAttribute), "System.ComponentModel.EditorBrowsableState.Never")]
// or even
[PropertyAttribute<EditorBrowsableAttribute>("System.ComponentModel.EditorBrowsableState.Never")]
```

This at least adds design-time checking that `EditorBrowsableAttribute` isn't typo'd, and means that doesn't have to be fully-qualified. However, constructor parameters still have to be fully-qualified, and it's starting to look a bit less like how the attribute will be written to the generated file, which makes grokking the feature a bit harder.

You could take this a step further and do:

```cs
[PropertyAttribute<EditorBrowsableAttribute>(EditorBrowsableState.Never)]
```

... which is probably an improvement. However, that falls down if you need to assign properties: I can't think of a nice way to let the user write the equivalent of:

```cs
[EditorBrowsable(State = EditorBrowsableState.Never)]
```

Another route is to let the user write a sort of proxy property to which attributes are applied, and those attributes are copied to the generated property. Something like:

```cs
[Notify] private int _bar;

[EditorBrowsable(EditorBrowsableState.Never)] private int BarProxy { get; set; }
```

This lets the user write attributes using normal attribute syntax on an actual property, which is nice. However, it does mean that there's this extra property floating around getting in the way, which is annoying and might interfere with whatever machinery is reading the attribute. You could get around that with an inner class I suppose:

```cs
[Notify] private int _bar;

private class AttributesProxy
{
    [EditorBrowsable(EditorBrowsableState.Never)] private int BarProxy { get; set; }
}
```

... but that's getting harder to grok again, and it places the attributes potentially a long way away from the point that they apply to, which makes things harder to maintain.

Any thoughts / preferences?

**Reply by canton7** [reply]
Another option is to let the user define a backing property, rather than a backing field. This is *probably* arleady supported -- I wrote most of the code with this in mind, but it's not particularly well tested. E.g.:

```cs
[EditorBrowsable(EditorBrowsableState.Never)]
[Notify] private int _foo { get; set; }
```

This gets around the issue of letter the user specify attributes which only apply to properties, not fields. However, it doesn't get around issues of the tooling looking for that property knowing to ignore the private backing property, even though it's got the attribute applied to it.

**Reply by canton7** [reply]
Yet another option is to allow something like:

```cs
[inpc: EditorBrowsable(EditorBrowsableState.Never)]
[Notify] private int _foo;
```

This raises:

> warning CS0658: 'inpc' is not a recognized attribute location. Valid attribute locations for this declaration are 'field, property'. All attributes in this block will be ignored.

... but that warning can be squashed (although it might annoy people with specific policies around suppressing warnings). Otherwise it works well: it's exposed nicely through Roslyn, but isn't applied to the field in the generated IL.

**Reply by hutterm** [reply]
Here's an idea:
could you generate an attribute with the same form as the real attribute in the same namespace as the real one, something like:

```csharp
    public class Test
    {
        [Notify]
        [Property_EditorBrowsable(EditorBrowsableState.Never)]
        private string _test;
    }
///generated:
namespace System.ComponentModel
{
    public class Property_EditorBrowsableAttribute : Attribute
    {
        public Property_EditorBrowsableAttribute(EditorBrowsableState state)
        {
            State = state;
        }
        public Property_EditorBrowsableAttribute() : this(EditorBrowsableState.Always)
        {  }
        public EditorBrowsableState State { get; }
    }
}
```
this would make it necessary to have the real namespace using.
not really sure how to discover the namespace though from the point of the generator

but it would allow to map all the same properties and make it look very much like the real attribute

**Reply by hutterm** [reply]
thinking more about this, I almost think that https://github.com/canton7/PropertyChanged.SourceGenerator/discussions/21#discussioncomment-3389189 would be the nicest and probably easiest option. Though It still has some magic to it and is not discoverable. Concerning naming: I'd almost go for naming it close to _notify_ so you have the connection to the `NotifyAttribute`. on the other hand `inpc:` is short...

**Reply by canton7** [reply]
Yeah, I like that it's neat and you get design-time support, and it mirrors how you'd apply an attribute to a compiler-generated backing field from a property. I don't like that we're forcing a warning onto the user - either they've got to add suppression around that line, or they've got to disable what is quite a useful warning across their whole project.

It's not entirely straightforward to implement either - we'll have to reconstruct the attribute from the parsed data which Roslyn gives us, making sure that everything is fully-qualified. Thankfully the range of ways you can write an attribute is fairly small.

I'm leaning towards the second option tbh. It's clunky, but it's obvious what's going on. Whatever string you use is pasted into the generated code verbatim. Simple. No warnings, no magic.

**Reply by hutterm** [reply]
Yeah, I think I agree. It would be straightforward.

**Comment by canton7** [comment, marked as answer]
I've released this as v1.0.5, [see the README](https://github.com/canton7/PropertyChanged.SourceGenerator#attributes-on-generated-properties). Let me know if you have any issues!
